src/datasets/features/features.py

```diff
@@ -443,17 +450,22 @@ def to_list_of_numpy(self, zero_copy_only=True):
         for dim in range(1, ndims):
             assert shape[dim] is not None, f"Support only dynamic size on first dimension. Got: {shape}"
 
+        null_indices = np.arange(len(storage))[storage.is_null().to_numpy(zero_copy_only=False)]
+
         arrays = []
         first_dim_offsets = np.array([off.as_py() for off in storage.offsets])
         for i in range(len(storage)):
-            storage_el = storage[i : i + 1]
-            first_dim = first_dim_offsets[i + 1] - first_dim_offsets[i]
-            # flatten storage
-            for _ in range(ndims):
-                storage_el = storage_el.flatten()
+            if i in null_indices:
+                arrays.append(np.nan)
```

Comment:
Nit: maybe we can optimize this. If the number of null indices is high this can take time:
```python
for i, is_null in enumerate(storage.is_null().to_numpy(zero_copy_only=False)):
    if is_null:
        arrays.append(np.nan)
    else:
        ...

Reply:
Great catch!